### Word Embedding Techniques using Embedding Layer in Keras

In [ ]:
### Libraries USed Tensorflow> 2.0  and keras

In [1]:
!pip install tensorflow-gpu

  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  Preparing metadata (setup.py) ... error
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.


In [2]:
import tensorflow as tf
print(tf.__version__)

2.15.0


In [3]:
##tensorflow >2.0
from tensorflow.keras.preprocessing.text import one_hot

In [4]:
### sentences
sent=[  'the glass of milk',
     'the glass of juice',
     'the cup of tea',
    'I am a good boy',
     'I am a good developer',
     'understand the meaning of words',
     'your videos are good']

In [5]:
sent

['the glass of milk',
 'the glass of juice',
 'the cup of tea',
 'I am a good boy',
 'I am a good developer',
 'understand the meaning of words',
 'your videos are good']

In [6]:
### Vocabulary size
voc_size=10000

#### One Hot Representation

In [8]:
onehot_repr=[one_hot(words,voc_size)for words in sent]
print(onehot_repr)

[[1810, 2189, 8676, 6707], [1810, 2189, 8676, 5641], [1810, 264, 8676, 3075], [7769, 8288, 5095, 8863, 4351], [7769, 8288, 5095, 8863, 1190], [5821, 1810, 2137, 8676, 2975], [5203, 7522, 7501, 8863]]


Index of these words in the vocablary           
the = 1810                                      
glass = 2189                                             
of = 8676                                         
milk = 6707                                     
The length of each one hot encoded representation is different

### Word Embedding Represntation

In [9]:
from tensorflow.keras.layers import Embedding
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential

In [10]:
import numpy as np

Since length is different, we use Padding
Padding: All the vectors with will converted to vec of same lengths          
Pre-Padding : add 0's before                 
Post-Padding: add 0's after                      

In [11]:
## pre padding
sent_length=8
embedded_docs=pad_sequences(onehot_repr,padding='pre',maxlen=sent_length)
print(embedded_docs)

[[   0    0    0    0 1810 2189 8676 6707]
 [   0    0    0    0 1810 2189 8676 5641]
 [   0    0    0    0 1810  264 8676 3075]
 [   0    0    0 7769 8288 5095 8863 4351]
 [   0    0    0 7769 8288 5095 8863 1190]
 [   0    0    0 5821 1810 2137 8676 2975]
 [   0    0    0    0 5203 7522 7501 8863]]


In [13]:
## 10 feature dimesnions
dim=10


In [14]:
model=Sequential()
model.add(Embedding(voc_size,10,input_length=sent_length))
model.compile('adam','mse')

In [15]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, 8, 10)             100000    
                                                                 
Total params: 100000 (390.62 KB)
Trainable params: 100000 (390.62 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [16]:
##'the glass of milk',
embedded_docs[0]

array([   0,    0,    0,    0, 1810, 2189, 8676, 6707], dtype=int32)

In [17]:
model.predict(embedded_docs[0])

1/1 [==============================] - 0s 115ms/step


array([[ 0.0308565 ,  0.03982483, -0.02580206, -0.02174062, -0.00244244,
         0.01799867,  0.00308079,  0.04977972, -0.02635801,  0.00246835],
       [ 0.0308565 ,  0.03982483, -0.02580206, -0.02174062, -0.00244244,
         0.01799867,  0.00308079,  0.04977972, -0.02635801,  0.00246835],
       [ 0.0308565 ,  0.03982483, -0.02580206, -0.02174062, -0.00244244,
         0.01799867,  0.00308079,  0.04977972, -0.02635801,  0.00246835],
       [ 0.0308565 ,  0.03982483, -0.02580206, -0.02174062, -0.00244244,
         0.01799867,  0.00308079,  0.04977972, -0.02635801,  0.00246835],
       [ 0.04108595, -0.0221459 ,  0.04308856,  0.04739568,  0.03680736,
         0.03719309, -0.02266966,  0.04161551,  0.0110934 , -0.02898753],
       [-0.04480777, -0.02635014,  0.04209166,  0.03727214,  0.00342407,
         0.04578735,  0.02306142,  0.04100529,  0.01644159,  0.02198717],
       [ 0.03672245,  0.03376115,  0.03822133,  0.03569874,  0.04336048,
         0.04140604, -0.0190938 ,  0.04319039

All the vector values in embedded_docs[0] are represented by 10 dimensions and the whole vector is represented by sentence length of 8

In [18]:
print(model.predict(embedded_docs))

1/1 [==============================] - 0s 35ms/step
[[[ 0.0308565   0.03982483 -0.02580206 -0.02174062 -0.00244244
    0.01799867  0.00308079  0.04977972 -0.02635801  0.00246835]
  [ 0.0308565   0.03982483 -0.02580206 -0.02174062 -0.00244244
    0.01799867  0.00308079  0.04977972 -0.02635801  0.00246835]
  [ 0.0308565   0.03982483 -0.02580206 -0.02174062 -0.00244244
    0.01799867  0.00308079  0.04977972 -0.02635801  0.00246835]
  [ 0.0308565   0.03982483 -0.02580206 -0.02174062 -0.00244244
    0.01799867  0.00308079  0.04977972 -0.02635801  0.00246835]
  [ 0.04108595 -0.0221459   0.04308856  0.04739568  0.03680736
    0.03719309 -0.02266966  0.04161551  0.0110934  -0.02898753]
  [-0.04480777 -0.02635014  0.04209166  0.03727214  0.00342407
    0.04578735  0.02306142  0.04100529  0.01644159  0.02198717]
  [ 0.03672245  0.03376115  0.03822133  0.03569874  0.04336048
    0.04140604 -0.0190938   0.04319039  0.04963927 -0.0091884 ]
  [ 0.0471532  -0.04623225 -0.04963659 -0.00149538 -0.02156

In [ ]:
### Assignment

sent=["The world is a better place",
      "Marvel series is my favourite movie",
      "I like DC movies",
      "the cat is eating the food",
      "Tom and Jerry is my favourite movie",
      "Python is my favourite programming language"
      ]